src/docregistry/registry.ts

```diff
@@ -717,24 +508,438 @@ class DocumentRegistry implements IDocumentRegistry {
 export
 namespace DocumentRegistry {
   /**
+   * The interface for a document model.
+   */
+  export
+  interface IModel extends IDisposable {
+    /**
+     * A signal emitted when the document content changes.
+     */
+    contentChanged: ISignal<this, void>;
+
+    /**
+     * A signal emitted when the model state changes.
+     */
+    stateChanged: ISignal<this, IChangedArgsGeneric<any>>;
+
+    /**
+     * The dirty state of the model.
+     *
+     * #### Notes
+     * This should be cleared when the document is loaded from
+     * or saved to disk.
+     */
+    dirty: boolean;
+
+    /**
+     * The read-only state of the model.
+     */
+    readOnly: boolean;
+
+    /**
+     * The default kernel name of the document.
+     */
+    readonly defaultKernelName: string;
+
+    /**
+     * The default kernel language of the document.
+     */
+    readonly defaultKernelLanguage: string;
+
+    /**
+     * Serialize the model to a string.
+     */
+    toString(): string;
+
+    /**
+     * Deserialize the model from a string.
+     *
+     * #### Notes
+     * Should emit a [contentChanged] signal.
+     */
+    fromString(value: string): void;
+
+    /**
+     * Serialize the model to JSON.
+     */
+    toJSON(): any;
+
+    /**
+     * Deserialize the model from JSON.
+     *
+     * #### Notes
+     * Should emit a [contentChanged] signal.
+     */
+    fromJSON(value: any): void;
+  }
+
+  /**
+   * The document context object.
+   */
+  export
+  interface IContext<T extends IModel> extends IDisposable {
+    /**
+     * A signal emitted when the kernel changes.
+     */
+    kernelChanged: ISignal<this, IKernel>;
```

Comment:
Do we want to start adding `readonly` to all signals? I see it happening in some spots of code that have been recently refactored.